pkg/planner/core/point_get_plan.go

```diff
@@ -339,16 +339,70 @@ func (p *PointGetPlan) LoadTableStats(ctx sessionctx.Context) {
 	loadTableStats(ctx, p.TblInfo, tableID)
 }
 
+// needsPartitionPruning checks if IndexValues can be used by GetPartitionIdxByRow() or if they have already been
+// converted to SortKey and would need GetPartitionIdxByRow() to be refactored to work, since it will unconditionally
+// convert it again.
+// Returns:
+// Matching partition
+// if done Partition pruning (else not needed, can use GetPartitionIdxByRow() instead)
+// error
+// TODO: Also supporting BatchPointGet? Problem is that partition ID must be mapped to handle/IndexValue.
+func needsPartitionPruning(sctx sessionctx.Context, tblInfo *model.TableInfo, pt table.PartitionedTable, dbName string, indexInfo *model.IndexInfo, indexCols []*expression.Column, indexValues []types.Datum, conds []expression.Expression, partitionNames []ast.CIStr) ([]int, bool, error) {
+	for i := range indexValues {
+		if tblInfo.Columns[indexInfo.Columns[i].Offset].FieldType.EvalType() != types.ETString ||
+			indexValues[i].Collation() == tblInfo.Columns[indexInfo.Columns[i].Offset].GetCollate() {
+			return nil, false, nil
+		}
+	}
+	// convertToPointGet will have the IndexValues already converted to SortKey,
+	// which will be converted again by GetPartitionIdxByRow, so we need to re-run the pruner
+	// with the conditions.
+
+	// TODO: Is there a simpler way, or existing function for this?!?
+	tblCols := make([]*expression.Column, 0, len(indexInfo.Columns))
+	var partNameSlice types.NameSlice
+	for _, tblCol := range tblInfo.Columns {
+		found := false
+		for _, idxCol := range indexCols {
+			if idxCol.ID == tblCol.ID {
+				tblCols = append(tblCols, idxCol)
+				found = true
+				break
+			}
+		}
+		partNameSlice = append(partNameSlice, &types.FieldName{
+			ColName:     tblCol.Name,
+			TblName:     tblInfo.Name,
+			DBName:      ast.NewCIStr(dbName),
+			OrigTblName: tblInfo.Name,
+			OrigColName: tblCol.Name,
+		})
+		if !found {
+			tblCols = append(tblCols, &expression.Column{
+				ID:       tblCol.ID,
+				OrigName: tblCol.Name.O,
+				RetType:  tblCol.FieldType.Clone(),
+			})
+		}
+	}
```

Comment:
Is this operations common for partitioning tables? If Yes, could we wrap this operation as a separate function in `partition_prune.go`?

Reply:
I tried using [`ColumnInfos2ColumnsAndNames()`](https://github.com/pingcap/tidb/blob/9d1dd4d3b0ff3dfae044ae6d4e300d64cd6adbdb/pkg/expression/expression.go#L1080) but there were [test failures](https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fghpr_unit_test/detail/ghpr_unit_test/28841/pipeline). So I'm keeping this code for now.